Sources/SwiftASN1/Basic ASN1 Types/PEMDocument.swift

```diff
@@ -199,32 +176,174 @@ public struct PEMDocument: Hashable, Sendable {
+            }
+
+            pemDocuments.append(try lazyPEMDocument.decode())
+        }
     }
+}
 
-    fileprivate var pemEndDiscriminator: String? {
-        return self.pemDiscriminator(expectedPrefix: "-----END ", expectedSuffix: "-----")
-    }
+/// A PEM document that has not yet decoded the base64 string found between the start and end marker.
+struct LazyPEMDocument {
+    var discriminator: Substring.UTF8View
+    var base64EncodedDERString: Substring.UTF8View
 
-    private func pemDiscriminator(expectedPrefix: String, expectedSuffix: String) -> String? {
-        var utf8Bytes = self.utf8[...]
+    func decode() throws -> PEMDocument {
+        guard let base64EncodedDERString = String(base64EncodedDERString) else {
+            throw ASN1Error.invalidPEMDocument(reason: "base64EncodedDERString is not valid UTF-8")
+        }
 
-        // We want to split this sequence into three parts: the prefix, the middle, and the end
-        let prefixSize = expectedPrefix.utf8.count
-        let suffixSize = expectedSuffix.utf8.count
+        guard let data = Data(base64Encoded: base64EncodedDERString, options: .ignoreUnknownCharacters) else {
+            throw ASN1Error.invalidPEMDocument(reason: "PEMDocument not correctly base64 encoded")
+        }
+        guard data.isEmpty == false else {
+            throw ASN1Error.invalidPEMDocument(reason: "PEMDocument has an empty body")
+        }
+        guard let type = String(discriminator) else {
+            throw ASN1Error.invalidPEMDocument(reason: "discriminator is not valid UTF-8")
+        }
 
-        let prefix = utf8Bytes.prefix(prefixSize)
-        utf8Bytes = utf8Bytes.dropFirst(prefixSize)
-        let suffix = utf8Bytes.suffix(suffixSize)
-        utf8Bytes = utf8Bytes.dropLast(suffixSize)
+        let derBytes = Array(data)
+        return PEMDocument(type: type, derBytes: derBytes)
+    }
+}
 
-        guard prefix.elementsEqual(expectedPrefix.utf8), suffix.elementsEqual(expectedSuffix.utf8) else {
+extension Substring.UTF8View {
+    /// A PEM document looks like this:
+    /// ```
+    /// -----BEGIN <SOME DISCRIMINATOR>-----
+    /// <base64 encoded bytes, 64 characters per line>
+    /// -----END <SOME DISCRIMINATOR>-----
+    /// ```
+    /// This function attempts find the BEGIN and END marker.
+    /// It then tries to extract the discriminator and the base64 encoded.
+    /// - Returns: `discriminator` found after BEGIN and END markers and `base64EncodedDerBytes`.
+    /// The `base64EncodedDerBytes` is as found in the original string and will still contain new lines if present in the original.
+    fileprivate mutating func readNextPEMDocument() throws -> LazyPEMDocument? {
+        /// First find the BEGIN marker: `-----BEGIN <SOME DISCRIMINATOR>-----
+        guard
+            let beginDiscriminatorPrefix = self.firstRange(of: "-----BEGIN ".utf8[...]),
+            let beginDiscriminatorSuffix = self[beginDiscriminatorPrefix.upperBound...].firstRange(
```

Comment:
Can we encapsulate this pattern in a helper function? We do it twice.

Reply:
Also this function could provide a few ranges, instead of the two we have here, so we can do more semantic named range lookups later in the function.

Reply:
have refactored it into this method:
```swift
extension Substring.UTF8View {
    func firstRangesOf(
        `prefix`: Substring,
        suffix: Substring
    ) -> (
        prefix: Range<Index>,
        infix: Range<Index>,
        suffix: Range<Index>
    )? 
}
```